crypto/ctype.c

```diff
@@ -0,0 +1,255 @@
+   /* 6B k */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 6C l */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 6D m */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 6E n */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 6F o */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 70 p */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 71 q */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 72 r */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 73 s */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 74 t */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 75 u */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 76 v */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 77 w */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 78 x */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 79 y */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 7A z */ CTYPE_MASK_graph | CTYPE_MASK_lower | CTYPE_MASK_print
+              | CTYPE_MASK_base64 | CTYPE_MASK_asn1print,
+   /* 7B { */ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 7C | */ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 7D } */ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 7E ~ */ CTYPE_MASK_graph | CTYPE_MASK_print | CTYPE_MASK_punct,
+   /* 7F   */ CTYPE_MASK_cntrl
+};
+
+int ossl_ctype_check(int c, unsigned int mask)
+{
+    const int max = OSSL_NELEM(ctype_char_map);
+
+#ifdef CHARSET_EBCDIC
+    /* Handle signed characters */
+    if (c < 0) {
+        if (c < -128)
+            return 0;
+        c += 256;
+    }
+    c = os_toascii[c];
+#endif
+    return c >= 0 && c < max && (ctype_char_map[c] & mask) != 0;
+}
+
+int ossl_ctype_ascii_check(int c, unsigned int mask)
+{
+    const int max = OSSL_NELEM(ctype_char_map);
+
+    return c >= 0 && c < max && (ctype_char_map[c] & mask) != 0;
+}
+
+#ifdef CHARSET_EBCDIC
+static const int case_change = 0x40;
```

Comment:
In other words suggestion is to `#if defined(CHARSET_EBCDIC) && !defined(CHARSET_EBCDIC_TEST)`